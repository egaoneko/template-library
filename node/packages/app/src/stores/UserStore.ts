import { IUser } from '@my-app/core/lib/interfaces/user';
import { makeAutoObservable } from 'mobx';
import { useMemo } from 'react';

class UserStore {
  public user: IUser | null = null;

  constructor(user: IUser | null) {
    makeAutoObservable(this);
    void this.clear();
    void this.hydrate(user);
  }

  public async hydrate(user: IUser | null): Promise<void> {
    await this.setUser(user);
  }

  private async clear(): Promise<void> {
    await this.setUser(null);
  }

  private async setUser(user: IUser | null): Promise<void> {
    this.user = user;
  }
}

export default UserStore;

let store: UserStore;

function initializeStore(user: IUser | null) {
  const _store = store ?? new UserStore(user);

  if (store?.user?.id !== _store.user?.id) {
    store = _store;
  }

  return _store;
}

export function useUserStore(user: IUser | null): UserStore {
  return useMemo(() => initializeStore(user), [user]);
}
