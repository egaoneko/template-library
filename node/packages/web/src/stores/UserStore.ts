import UserAPI from '@api/user';
import { makeAutoObservable } from 'mobx';
import { IS_SSR } from '@constants/common';
import { useMemo } from 'react';
import { IUser, LoginRequest, RegisterRequest, UpdateRequest } from '@interfaces/user';
import AuthAPI from '@api/auth';
import { notifyError, notifySuccess } from '@utils/notifiy';
import { removeToken, setToken } from '@utils/cookie';

export class UserStore {
  public lastUpdate: number = 0;
  public user: IUser | null = null;

  constructor(user: IUser | null) {
    makeAutoObservable(this);
    void this.clear();
    void this.hydrate(user);
  }

  public async hydrate(user: IUser | null): Promise<void> {
    await this.setUser(user);
  }

  public async register(request: RegisterRequest): Promise<boolean> {
    try {
      const user = await AuthAPI.register(request);
      notifySuccess('Successfully registered!');
      return !!user;
    } catch (e) {
      notifyError(e.response?.data?.message ?? e.message);
    }

    return false;
  }

  public async login(request: LoginRequest): Promise<IUser | null> {
    try {
      const user = await AuthAPI.login(request);
      this.setUser(user);

      if (this.user) {
        notifySuccess('Successfully login!');
      } else {
        notifyError('Fail to login');
      }
    } catch (e) {
      notifyError(e.response?.data?.message ?? e.message);
    }

    return this.user;
  }

  public async update(request: UpdateRequest): Promise<IUser | null> {
    try {
      const user = await UserAPI.update(request);
      this.setUser({
        ...this.user,
        ...user,
      });

      if (this.user) {
        notifySuccess('Successfully updated!');
      } else {
        notifyError('Fail to update');
      }
    } catch (e) {
      notifyError(e.response?.data?.message ?? e.message);
    }

    return this.user;
  }

  public async logout(): Promise<void> {
    try {
      await AuthAPI.logout();
      removeToken();
      await this.clear();
    } catch (e) {
      notifyError(e.message);
    }
  }

  private async clear(): Promise<void> {
    await this.setUser(null);
  }

  private async setUser(user: IUser | null): Promise<void> {
    this.user = user;

    if (!this.user || !this.user?.token) {
      return;
    }

    setToken(this.user.token, this.user.refreshToken);
  }
}

let store: UserStore;

function initializeStore(user: IUser | null) {
  const _store = store ?? new UserStore(user);

  if (IS_SSR) {
    return _store;
  }

  if (store?.user?.id !== _store.user?.id) {
    store = _store;
  }

  return _store;
}

export function useUserStore(user: IUser | null): UserStore {
  return useMemo(() => initializeStore(user), [user?.id]);
}
