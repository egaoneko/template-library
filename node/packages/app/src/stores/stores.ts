import { useContext } from 'react';
import { MobXProviderContext } from 'mobx-react';

import UserStore from './UserStore';

export interface Stores {
  userStore: UserStore;
}

export function useStores(): Stores {
  return useContext(MobXProviderContext) as Stores;
}
