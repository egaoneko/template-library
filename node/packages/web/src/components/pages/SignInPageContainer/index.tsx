import HeaderTemplates from '@components/templates/layout/HeaderTemplates';
import Head from '@components/atoms/layout/Head';
import React, { FC, ReactNode, useState } from 'react';
import SignInContentTemplate from './templates/SignInContentTemplate';
import { LoginRequest } from '@interfaces/user';
import { useRouter } from 'next/router';
import { useStores } from '@stores/stores';

interface PropsType {
  children?: ReactNode;
}

const SignInPageContainer: FC<PropsType> = props => {
  const { userStore } = useStores();
  const router = useRouter();
  const [loading, setLoading] = useState<boolean>(false);

  function onFinish(request: LoginRequest): Promise<void> {
    setLoading(true);
    return userStore.login(request).then(user => {
      if (!user) {
        setLoading(false);
        return;
      }
      router.push('/');
    });
  }

  return (
    <>
      <Head title={'LOGIN'} />
      <HeaderTemplates headingTitle={'conduit'} />
      <SignInContentTemplate loading={loading} onFinish={onFinish} />
    </>
  );
};

export default SignInPageContainer;
