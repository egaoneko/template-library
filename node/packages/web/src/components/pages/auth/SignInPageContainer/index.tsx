import React, { FC, useState } from 'react';
import { LoginRequest } from '@my-app/core/lib/interfaces/user';
import { useRouter } from 'next/router';
import { BasePropsType } from '@my-app/core/lib/interfaces/common';

import { useStores } from 'src/stores/stores';
import Head from 'src/components/atoms/layout/Head';
import BaseLayoutTemplate from 'src/components/templates/layout/BaseLayoutTemplate';

import SignInContentTemplate from './templates/SignInContentTemplate';

interface PropsType extends BasePropsType {
  successUrl?: string;
}

const SignInPageContainer: FC<PropsType> = props => {
  const { userStore } = useStores();
  const router = useRouter();
  const [loading, setLoading] = useState<boolean>(false);

  const handleOnFinish = async (request: LoginRequest): Promise<void> => {
    setLoading(true);
    return userStore.login(request).then(async user => {
      if (!user) {
        setLoading(false);
        return;
      }

      if (props.successUrl) {
        await router.push(props.successUrl);
      } else {
        await router.push('/');
      }
    });
  };

  return (
    <BaseLayoutTemplate pathname={props.pathname}>
      <Head title={'LOGIN'} />
      <SignInContentTemplate loading={loading} onFinish={handleOnFinish} />
    </BaseLayoutTemplate>
  );
};

export default SignInPageContainer;
