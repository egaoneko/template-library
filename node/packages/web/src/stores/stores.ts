import { useContext } from 'react';
import { enableStaticRendering, MobXProviderContext } from 'mobx-react';
import { IS_SSR } from 'src/constants/common';
import { UserStore } from './UserStore';

enableStaticRendering(IS_SSR);

export interface Stores {
  userStore: UserStore;
}

export function useStores(): Stores {
  return useContext(MobXProviderContext) as Stores;
}
