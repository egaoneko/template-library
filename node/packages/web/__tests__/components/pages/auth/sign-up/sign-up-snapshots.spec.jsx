import React from 'react';
import renderer from 'react-test-renderer';
import { UserStore } from '../../../../../src/stores/UserStore';
import { Provider } from 'mobx-react';
import SignInPageContainer from '../../../../../src/components/pages/auth/SignUpPageContainer';

describe('Sign up Snapshots', () => {
  let userStore;

  beforeEach(() => {
    userStore = new UserStore();
  });

  it('renders sign up', () => {
    const tree = renderer
      .create(
        <Provider userStore={userStore}>
          <SignInPageContainer />
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
