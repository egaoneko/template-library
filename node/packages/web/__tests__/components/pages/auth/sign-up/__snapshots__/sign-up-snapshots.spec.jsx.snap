// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign up Snapshots renders sign up 1`] = `
.c4 {
  width: 100%;
  position: absolute;
  bottom: 0px;
  margin-top: 3rem;
  --tw-bg-opacity: 1;
  background-color: rgba(229,231,235,var(--tw-bg-opacity));
  padding-left: 3.5rem;
  padding-right: 3.5rem;
  padding-top: 1rem;
  padding-bottom: 1rem;
  font-size: 0.75rem;
  line-height: 1rem;
  --tw-text-opacity: 1;
  color: rgba(156,163,175,var(--tw-text-opacity));
}

.c5 {
  color: var(--color-primary);
}

.c5:hover {
  color: var(--color-secondary);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c1 {
  width: 100%;
  margin-left: auto;
  margin-right: auto;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  padding-top: 0.5rem;
  padding-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 {
  font-weight: 700;
  font-size: 1.5rem;
  line-height: 2rem;
  color: var(--color-primary);
  font-family: ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,"Segoe UI",Roboto,"Helvetica Neue",Arial,"Noto Sans",sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol","Noto Color Emoji";
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 {
  width: 100%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

@media (min-width:640px) {
  .c1 {
    max-width: 640px;
  }
}

@media (min-width:768px) {
  .c1 {
    max-width: 768px;
  }
}

@media (min-width:1024px) {
  .c1 {
    max-width: 1024px;
  }
}

@media (min-width:1280px) {
  .c1 {
    max-width: 1280px;
  }
}

@media (min-width:1536px) {
  .c1 {
    max-width: 1536px;
  }
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <div
      className="c2"
    >
      <div
        className="c3"
        data-cy="header-logo-link"
      >
        <a
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          conduit
        </a>
      </div>
      <div
        className="text-base leading-6 text-gray-400 font-sans flex items-center"
      >
        <div
          className="py-2 ml-4 cursor-pointer first:ml-0 hover:text-gray-800 flex "
          data-cy="header-home-link"
        >
          <div
            className="flex items-center"
            href="/"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span
              className="mr-1"
            />
            <span>
              Home
            </span>
          </div>
        </div>
        <div
          className="py-2 ml-4 cursor-pointer first:ml-0 hover:text-gray-800 flex "
          data-cy="header-sign-in-link"
        >
          <div
            className="flex items-center"
            href="/auth/sign-in"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span
              className="mr-1"
            />
            <span>
              Sign in
            </span>
          </div>
        </div>
        <div
          className="py-2 ml-4 cursor-pointer first:ml-0 hover:text-gray-800 flex "
          data-cy="header-sign-up-link"
        >
          <div
            className="flex items-center"
            href="/auth/sign-up"
            onClick={[Function]}
            onMouseEnter={[Function]}
          >
            <span
              className="mr-1"
            />
            <span>
              Sign up
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="container mx-auto mt-8"
  >
    <div
      className="w-full max-w-md mx-auto"
    >
      <h1
        className="text-4xl text-center mb-4"
        data-cy="content-page-title"
      >
        Sign Up
      </h1>
      <div
        className="text-primary hover:text-secondary text-center mb-4"
        data-cy="content-sign-in-link"
      >
        <a
          href="/auth/sign-in"
          onClick={[Function]}
          onMouseEnter={[Function]}
        >
          Have an account?
        </a>
      </div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="flex flex-col gap-4"
        >
          <div>
            <input
              className=" block w-full rounded-md border-gray-300 shadow-sm focus:border-indigo-300 focus:ring focus:ring-indigo-200 focus:ring-opacity-50 disabled:opacity-50 disabled:cursor-not-allowed"
              data-cy="content-form-input-username"
              disabled={false}
              name="username"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Username"
              type="text"
            />
          </div>
          <div>
            <input
              className=" block w-full rounded-md border-gray-300 shadow-sm focus:border-indigo-300 focus:ring focus:ring-indigo-200 focus:ring-opacity-50 disabled:opacity-50 disabled:cursor-not-allowed"
              data-cy="content-form-input-email"
              disabled={false}
              name="email"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Email"
              type="email"
            />
          </div>
          <div>
            <input
              className=" block w-full rounded-md border-gray-300 shadow-sm focus:border-indigo-300 focus:ring focus:ring-indigo-200 focus:ring-opacity-50 disabled:opacity-50 disabled:cursor-not-allowed"
              data-cy="content-form-input-password"
              disabled={false}
              name="password"
              onBlur={[Function]}
              onChange={[Function]}
              placeholder="Password"
              type="password"
            />
          </div>
          <button
            className="ml-auto px-6 py-3 hover:text-opacity-80 hover:bg-opacity-80 hover:border-opacity-80 font-bold border rounded focus:outline-none focus:shadow-outline disabled:opacity-50 disabled:cursor-not-allowed bg-primary border-primary text-white"
            data-cy="content-form-button-submit"
            disabled={false}
            type="submit"
          >
            Sing up
          </button>
        </div>
      </form>
    </div>
  </div>
  <div
    className="c4"
  >
    <span
      className="c5 text-base mr-1 font-bold"
      data-cy="footer-logo-link"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      conduit
    </span>
    Â© 2020. An interactive learning project from
    <span
      className="c5 ml-1"
      data-cy="footer-site-link"
      href="https://thinkster.io/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      Thinkster
    </span>
    . Code licensed under MIT.
  </div>
</div>
`;
